first draft of blog post

oscarbatori opened the pull request:
To do:

- [x] finalize technical details of examples
- [x] copy editing/possibly iterate on assets
- [x] incorporate feedback from Metaflow team

Commit:
first draft of blog post

Commit:
Updated the first half plus improved the assets

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
Put the exact number

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
put an actual query and results

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
make a hypothesis?

timsehn replied:
dataset?

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
hypothesis

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
Maybe a sentence about why this is important or if it's not, why it was chosen.

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
Missing a picture?

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
I really like this paragraph.

oscarbatori replied:
I ended up dropping it, but I might work back in some how as I agree it explains something cool

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
Should definitely have called it Meatflow

oscarbatori replied:
true... flows metaflow metadata data wtf

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
Now what?

timsehn replied:
A diff example would be good here.

oscarbatori replied:
dropped this, might reintroduce

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
I think you have to mention Lineage, Reproducibility, and Back Testing here. That's really what Metaflow users get.

timsehn commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,170 @@
+---
+
+## Background
+This post details how to use Metaflow with Dolt. [Metaflow](https://metaflow.org/) is a workflow manager that offers data scientists the ability to define small experiments and production jobs from a single API. [Dolt](https://docs.dolthub.com/) is a version controlled relational database. It provides a familiar SQL interface along with Git-like version control features for capturing state. Each  ...
```
I think you need to mention in the opening that you want to use Dolt to get lineage, reproducibility, and back testing.

timsehn reviewed (commented):
B+. Much better than usual especially at avoiding word salad.

Commit:
Another draft of blog post

max-hoffman commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,164 @@
+## Background
+This post details how to use Metaflow with Dolt. [Metaflow](https://metaflow.org/) is a workflow manager that offers data scientists the ability to define local experiments and scale those experiments to production jobs from a single API. [Dolt](https://docs.dolthub.com/) is a version controlled relational database. It provides a familiar SQL interface along with Git-like version control features ...
+
+Using Dolt as a source and target for Metaflow jobs that users can reproduce, trace the lineage of, and employ complex back testing logic against.
```
make full sentence

max-hoffman commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
i can't make this syntax work. we could do `doltpy[metaflow]` maybe

max-hoffman commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
write -> right

max-hoffman commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
i like this addition, maybe simplify this bit "...very directly impacts how effective such efforts are as a means of collecting data."

max-hoffman commented on doltpy_integrations/metaflow/examples/hospital_bounty/blog_post.md:
4-space indenting flag lines is standard

max-hoffman committed:
Bump dolt version

Commit:
Updates:
- remove dependency on doltpy, replace with skinny doltcli
- implement some basic pandas read and writes on top of doltcli
- fix unit tests
- more updates to blog post

max-hoffman commented on dolt_integrations/metaflow/dolt.py:
```diff
@@ -258,7 +252,7 @@ def _execute_read_action(self, action: DoltAction, config: DoltConfig):
         starting_commit = self._get_latest_commit_hash(db)
         try:
-            db.sql(query=f"set `@@{db.repo_name}_head` = '{action.commit}'")
+            db.sql(query=f"set `@@{db.repo_name}_head` = '{action.commit}'", result_format='csv')
```
merge main, i think i fixed this

max-hoffman commented on tests/metaflow/test_metaflow.py:
```diff
@@ -66,7 +74,7 @@ def test_branchdt_cm_write(active_run, dolt_config, doltdb):
     db = Dolt(doltdb)
     output_df = read_pandas_sql(db, "SELECT * from `baz`")
-    np.testing.assert_array_equal(output_df.A.values, ["2","2","2"])
+    np.testing.assert_array_equal(output_df.A.values, [2, 2, 2])
```
nice, that was bothering me but i never got around to it

max-hoffman commented on pyproject.toml:
```diff
@@ -1,13 +1,14 @@
 [tool.poetry]
-name = "doltpy_integrations"
+name = "dolt_integrations"
```
this is a better name, good idea

max-hoffman commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,164 @@
+## Background
+This post details how to use Metaflow with Dolt. [Metaflow](https://metaflow.org/) is a workflow manager that offers data scientists the ability to define local experiments and scale those experiments to production jobs from a single API. [Dolt](https://docs.dolthub.com/) is a version controlled relational database. It provides a familiar SQL interface along with Git-like version control features ...
+
+Using Dolt as a source and target for Metaflow jobs that users can reproduce, trace the lineage of, and employ complex back testing logic against.
```
sentence fragment

max-hoffman commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
nit - "completely reproducible" might be redundant

max-hoffman commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
"get stuck some arm chair data science"

max-hoffman replied:
if you don't like the phrase you can drop. some way to suggest upfront that we aren't being rigorous with our analysis seems like a good idea

max-hoffman commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
nit -- indent

max-hoffman committed:
Pull conflicts

max-hoffman committed:
Fix hanging comma that was casting query string to tuple

max-hoffman committed:
Fix standalone inaactive run audit error bug

max-hoffman committed:
Black

Commit:
instantiate a DoltDT via a flow run path

Commit:
Blog post ready for review

max-hoffman commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
can you show a head or something? for the `dolt sql` below also, missing output makes my lizard brain unhappy

timsehn replied:
agreed

max-hoffman commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
again, missing outputs here makes me wonder if it actually works

max-hoffman commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,267 @@
+## Background
+This post details how to use Metaflow with Dolt. [Metaflow](https://metaflow.org/) is a workflow manager that offers data scientists the ability to define local experiments and scale those experiments to production jobs from a single API. [Dolt](https://docs.dolthub.com/) is a version controlled relational database. It provides a familiar SQL interface along with Git-like version control features ...
+
+To illustrate the power of this integration we use an example of a Metaflow piepline that produces a final result as a Pandas data-frame that is stored in a database for use by application layer services:
```
data-frame -> DataFrame

max-hoffman reviewed (approved):
LG, let's ship

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,267 @@
+---
+
+## Background
+This post details how to use Metaflow with Dolt. [Metaflow](https://metaflow.org/) is a workflow manager that offers data scientists the ability to define local experiments and scale those experiments to production jobs from a single API. [Dolt](https://docs.dolthub.com/) is a version controlled relational database. It provides a familiar SQL interface along with Git-like version control features ...
```
remove "for capturing historical states". It's too specific. Dolt does more than that.

timsehn replied:
"Each commit corresponds to a complete state" -> the complete state?

timsehn replied:
"Each commit corresponds to a complete state of the database at the time the commit was created." -> This sentence is weird. What are you trying to say? A commit does not represent a "time". Multiple commits can be made at the same time.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,267 @@
+## Background
+This post details how to use Metaflow with Dolt. [Metaflow](https://metaflow.org/) is a workflow manager that offers data scientists the ability to define local experiments and scale those experiments to production jobs from a single API. [Dolt](https://docs.dolthub.com/) is a version controlled relational database. It provides a familiar SQL interface along with Git-like version control features ...
+
+To illustrate the power of this integration we use an example of a Metaflow piepline that produces a final result as a Pandas data-frame that is stored in a database for use by application layer services:
```
"To illustrate the power of integrating Metaflow and Dolt,""

timsehn replied:
"We use an example Metaflow pipeline" -> lose the "of a"

timsehn replied:
"To illustrate the power of this integration we use an example of a Metaflow piepline that produces a final result as a Pandas data-frame that is stored in a database for use by application layer services" 

Should be something like:

"To illustrate the power of integrating Metaflow and Dolt, we use an example Metaflow piepline to produce a Pandas DataFrame stored in a Dolt database for use by application layer services."

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,267 @@
+To illustrate the power of this integration we use an example of a Metaflow piepline that produces a final result as a Pandas data-frame that is stored in a database for use by application layer services:
+![]()
+
+This blog post will breakdown how to use Dolt to store final and intermediate results to augment Metaflow based pipelines with full reproducibility, lineage, and back-testing capabilities. Dolt allows users to see where their data came from, what it looked like at every transformation, and to feed historical versions of it into new jobs.
```
remove "to store final and intermediate results"

timsehn replied:
if you want to say that put it in a different sentence. It complicates the lede.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
remove "when its run"

timsehn reviewed (changes requested):
Can one of you please go through and ruthlessly murder words? Every sentence has fluff I need to cut.

Commit:
murdered some words

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,270 @@
+---
+
+## Background
+This post details how to use Metaflow with Dolt. [Metaflow](https://metaflow.org/) is a workflow manager that offers data scientists the ability to define local experiments and scale those experiments to production jobs from a single API. [Dolt](https://docs.dolthub.com/) is a version controlled relational database. It provides a familiar SQL interface along with Git-like version control features ...
```
"complete state of the database at the time the commit was created" -> "when the commit was created"

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,270 @@
+## Background
+This post details how to use Metaflow with Dolt. [Metaflow](https://metaflow.org/) is a workflow manager that offers data scientists the ability to define local experiments and scale those experiments to production jobs from a single API. [Dolt](https://docs.dolthub.com/) is a version controlled relational database. It provides a familiar SQL interface along with Git-like version control features ...
+
+To illustrate the power of integrating Metaflow and Dolt, we use an example Metaflow pipeline to produce a Pandas DataFrame stored in a Dolt database for use by application layer services. The goal is for application layer code to read from datasets that are versioned and reproducible to create a more healthy boundary between data science production services using Dolt's version control features:
```
I know i recommended this sentence but I think it's better broken up. 

"To illustrate the power of integrating Metaflow and Dolt, we use an example Metaflow pipeline to derive results data from an input dataset. We store this results data temporarily in a Pandas DataFrame. We convert that DataFrame to permanent storage in Dolt for use with application layer services." 

Or something.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,270 @@
+To illustrate the power of integrating Metaflow and Dolt, we use an example Metaflow pipeline to produce a Pandas DataFrame stored in a Dolt database for use by application layer services. The goal is for application layer code to read from datasets that are versioned and reproducible to create a more healthy boundary between data science production services using Dolt's version control features:
+![Data Architecture](dolt-metaflow-integration-data-architecture.png)
+
+This blog post will breakdown how to use Dolt to augment Metaflow based pipelines with full reproducibility, lineage, and back-testing capabilities for tabular data read from and written to Dolt. Dolt allows users to see where their data came from, what it looked like at every transformation, and to feed historical versions of it into subsequent flow runs.
```
delete "read from and written to Dolt."

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,270 @@
+This blog post will breakdown how to use Dolt to augment Metaflow based pipelines with full reproducibility, lineage, and back-testing capabilities for tabular data read from and written to Dolt. Dolt allows users to see where their data came from, what it looked like at every transformation, and to feed historical versions of it into subsequent flow runs.
+
+## Example Pipeline
+Our pipeline consists of two flows. One flow consumes the results written by the other. The first flow computes the state level median price for a hospital procedure. The second flow computes the variance of the price for a procedure across states. 
```
Maybe start with:

For input data we chose a [new public dataset of 1400 US hospital prices](https://www.dolthub.com/repositories/dolthub/hospital-price-transparency) created using [DoltHub data bounties](https://www.dolthub.com/bounties).

It's jarring not knowing where the input data came from.

oscarbatori replied:
Good shout, I actually cut a section on bounties that explained this and didn't introduce something appropriate here.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
```diff
@@ -0,0 +1,270 @@
+We will use the end result to illustrate how integrating Dolt and Metaflow provides Metaflow users with the ability to traverse versions of their final data, as well as traceback through the pipeline to examine various stages. Users can do this via the Metaflow API and Dolt integration in a familiar environment like a console or a notebook.
+
+## How it Works
+Our design goal with this integration was to give the Metaflow user additional capabilities directly from Metaflow. We wanted to minimize the additional API surface area required.
```
"We wanted to minimize the additional API surface area required." -> "We wanted to minimize additional API surface area."

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
"Since Dolt database root hashes are unique," -> "Since Dolt commit hashes are unique,"

timsehn replied:
I might even drop it and just go with:

Each time a run interacts with Dolt, we capture a small amount of metadata to make that interaction reproducible.  We create a mapping between Metaflow runs and Dolt commits to provide users with powerful lineage and reproducibility features.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
"Runs of a flow read and write Pandas DataFrame objects to Dolt":

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
a remote Dolt database

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
This might need a why...

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
delete actually.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
of commit `gstcq5loi9ieqdv1elrljab9hcgr090p`, the first one labeled "Updating hospitals with changes from hospitals.csv.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
same here.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
variance's and median's

oscarbatori replied:
I don't think this is actually possessive since those are plurals

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
A diff would be really cool.

Commit:
copy edits, and example runs

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
````diff
@@ -0,0 +1,400 @@
++-----------+
+```
+
+Out the outset we said we were going to store the results in Dolt. To do so we need to create a Dolt database to write to:
````
At the,

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
````diff
@@ -0,0 +1,400 @@
+```
+Before recomputing the medians let's create a branch in our `hospital-price-analysis` database associated with the root commit to store these experiments:
+```
+
````
Why is this blank?

oscarbatori replied:
Thanks forgot to add.

timsehn commented on dolt_integrations/metaflow/examples/hospital_bounty/blog_post.md:
````diff
@@ -0,0 +1,400 @@
+print(df)
+```
+
+Since this commit was taken from far earlier in the data gathering process, we can see far fewer unique procedure codes:
````
Add, 21K vs 1.17M at the end.

timsehn reviewed (approved):
Much better. Passed my writing filter. A couple comments.

Commit:
all examples working, more copy editing

max-hoffman committed:
Add release stuff

Commit:
final copy changes
